src/GraphQl/Resolver/Factory/ItemResolverFactory.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\GraphQl\Resolver\Factory;
+
+use ApiPlatform\Core\Api\IriConverterInterface;
+use ApiPlatform\Core\Exception\ItemNotFoundException;
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\GraphQl\Resolver\FieldsToAttributesTrait;
+use ApiPlatform\Core\GraphQl\Resolver\QueryItemResolverInterface;
+use ApiPlatform\Core\GraphQl\Resolver\ResourceAccessCheckerTrait;
+use ApiPlatform\Core\GraphQl\Serializer\ItemNormalizer;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Security\ResourceAccessCheckerInterface;
+use ApiPlatform\Core\Util\ClassInfoTrait;
+use GraphQL\Type\Definition\ResolveInfo;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Creates a function retrieving an item to resolve a GraphQL query.
+ *
+ * @experimental
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class ItemResolverFactory implements ResolverFactoryInterface
+{
+    use ClassInfoTrait;
+    use FieldsToAttributesTrait;
+    use ResourceAccessCheckerTrait;
+
+    private $iriConverter;
+    private $queryResolverLocator;
+    private $resourceAccessChecker;
+    private $normalizer;
+    private $resourceMetadataFactory;
+
+    public function __construct(IriConverterInterface $iriConverter, ContainerInterface $queryResolverLocator, NormalizerInterface $normalizer, ResourceMetadataFactoryInterface $resourceMetadataFactory, ResourceAccessCheckerInterface $resourceAccessChecker = null)
+    {
+        $this->iriConverter = $iriConverter;
+        $this->queryResolverLocator = $queryResolverLocator;
+        $this->normalizer = $normalizer;
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+        $this->resourceAccessChecker = $resourceAccessChecker;
+    }
+
+    public function __invoke(string $resourceClass = null, string $rootClass = null, string $operationName = null): callable
+    {
+        return function ($source, $args, $context, ResolveInfo $info) use ($operationName) {
+            // Data already fetched and normalized (field or nested resource)
+            if (isset($source[$info->fieldName])) {
+                return $source[$info->fieldName];
+            }
+
+            if (!isset($args['id'])) {
```

Comment:
I think this check required only for bundled resolver. Initially, one of the reasons of introduction of custom operations was  an idea to create endpoints/resources without compulsory arguments. Probably `iri` should be grabbed from custom operation configuration in case of unavailability of `id`.

Reply:
I agree. I think moving normalization to the factory makes sense, but the retrieval of items should be handled by the resolvers.

`iri` should not be needed either, and the `$item` parameter should be removed from `QueryItemResolverInterface`. Maybe we can pass the `iriConverter` and `baseNormalizationContext`, or some callable preconfigured with those, to the resolver - so you can actually load items inside of it?

Reply:
I think there will be three ways to add custom operations in the future:
- custom queries (this one). `id` arg is mandatory, the custom resolver receives the item/collection, it does some things, it returns it.
- custom mutations: an `input` is mandatory, the custom resolver receives the denormalized item, it does some things, it returns it and it will be persisted afterwards.
- custom operations: nothing is mandatory for the args, the resolver does anything it wants, nothing is done if it returns something (it needs to do the normalization itself). I'm not sure about this one. And IMO it can be strange to define it on a resource since you can do anything you want.
WDYT?

Reply:
My example (and currently implemented use-case) for custom queries was a "currentUser" query (returns logged in user based on token storage). Having an `id` arg does not really make sense here, but using the resource normalization does.

Reply:
OK. I will do the modifications.

Reply:
Thank you! 🙂🚀

Reply:
What's the use-case for such custom queries (having `id`) then? It could be done by specifying `@ApiResource({grapqhl={"query"}})`. My comment mentioned in the starting message of PR is about resources free from `id`.

Reply:
Custom queries having an `id` argument could be useful to modify the received item or to add side effects to your query.
WIth what I've done, you have an `id` argument by default and you receive the item by default but you can easily change this behavior.

Reply:
I believe having the optional `id` argument should be fine for now 👍 If you don't need / use it, just ignore the item argument inside your resolver.